Mathlib/Util/Simp.lean

```diff
@@ -0,0 +1,34 @@
+/-
+Copyright (c) 2025 Paul Lezeau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau
+-/
+import Lean.Meta.Tactic.Simp.Types
+import Qq
+
+/-! # Additional simp utilities
+
+This file adds additional tools for metaprogramming with the `simp` tactic
+
+-/
+
+open Lean Meta Qq
+
+namespace Simp
+
+open Simp
+
+/-- `Qq` version of `Simp.Methods.discharge?`, which avoids having to use `~q` matching
+on the proof expression returned by `Methods.discharge?`
+
+`Methods.dischargeQ? (a : Q(Prop))` attempts to prove `a` using the discharger, returning
+`some (pf : Q(a))` if a proof is found and `none` otherwise.
+-/
+def Methods.dischargeQ? (M : Methods) (a : Q(Prop)) : SimpM <| Option Q($a) := do
+  match ← M.discharge? a with
+  | some pf =>
+    let ⟨0, ~q($a), pf⟩ ← inferTypeQ pf | return none
```

Comment:
I think if this has the wrong type you want to report an error, not silently return `none`.

One way you could do that is
```suggestion
    let pf : Q($a) := pf
    pf.check
    return pf
```